File: crates/kuben/src/cli/upgrade.rs
//! Upgrades (M4.8): `kuben upgrade-check`, run with the *new* binary before
//! it replaces the old one, and the journaled migration every start and
//! `kuben migrate` go through, with a backup first when migrations are
//! pending.

use std::{path::Path, process::Command};

use kuben_core::{
    config::Config,
    time::now_ms,
    upgrade::{Facts, Level, preflight},
};
use kuben_store::Store;

use crate::cli::{UpgradeCheckOpts, VERSION};

/// Free bytes of the file system holding `dir` (its nearest existing
/// ancestor), from `df`.
fn free_bytes(dir: &Path) -> Option<u64> {
    let existing = dir.ancestors().find(|p| p.exists())?;
    let out = Command::new("df").arg("-Pk").arg(existing).output().ok()?;
    parse_df(&String::from_utf8_lossy(&out.stdout))
}

/// The available KiB of the last line of `df -Pk`, in bytes.
fn parse_df(output: &str) -> Option<u64> {
    let fields: Vec<&str> = output.lines().last()?.split_whitespace().collect();
    fields.get(3)?.parse::<u64>().ok().map(|kib| kib << 10)
}

/// What the preflight of this binary against `store` sees.
async fn facts(cfg: &Config, store: &Store, major: bool) -> anyhow::Result<Facts> {
    let schema = store.schema_state().await?;
    let versions = store.server_versions().await?;
    let last_backup = if schema.applied >= 25 {
        store.last_backup().await?.map(|b| b.finished_at)
    } else {
        None
    };
    Ok(Facts {
        running: versions.into_iter().next(),
        target: VERSION.to_owned(),
        major,
        schema: schema.applied,
        target_schema: Store::latest_migration(),
        dirty_schema: schema.dirty,
        last_backup,
        now: now_ms(),
        active_operations: if schema.applied > 0 {
            store.active_operations().await?
        } else {
            0
        },
        agents: if schema.applied > 0 {
            store.agent_protocols().await?
        } else {
            Vec::new()
        },
        target_protocols: kuben_agent::protocol::SUPPORTED_VERSIONS,
        free_bytes: free_bytes(&cfg.backup_dir()),
    })
}

/// `kuben upgrade-check`: every finding; fails when any does.
pub async fn check(cfg: Config, opts: UpgradeCheckOpts) -> anyhow::Result<()> {
    let store = Store::connect_unmigrated(&cfg.database).await?;
    let findings = preflight(&facts(&cfg, &store, opts.major).await?);
    for f in &findings {
        let mark = match f.level {
            Level::Ok => "OK  ",
            Level::Warn => "WARN",
            Level::Fail => "FAIL",
        };
        println!("{mark}  {:<11} {}", f.check, f.detail);
    }
    store.close().await?;
    if findings.iter().any(|f| f.level == Level::Fail) {
        anyhow::bail!("this upgrade is not safe yet: fix the failures above, nothing has changed");
    }
    println!("ready to upgrade to {VERSION}");
    Ok(())
}

/// Migrate `store` as this version, journaled; when migrations are pending
/// on a database with data, take a backup first where PostgreSQL's client
/// tools are installed. A failed backup stops the migration.
pub async fn migrate(cfg: &Config, store: &Store) -> anyhow::Result<()> {
    let state = store.guard_schema().await?;
    let pending = Store::latest_migration() - state.applied;
    if pending > 0 && state.applied > 0 && cfg.backup.before_upgrade {
        if crate::cli::backup::client_tools_installed() {
            let opts = crate::cli::BackupOpts {
                out: None,
                include_keyring: false,
                keep: None,
                scheduled: true,
            };
            tracing::info!(pending, "taking a backup before the migrations");
            crate::cli::backup::run(cfg.clone(), opts).await?;
        } else {
            tracing::warn!(
                pending,
                "migrations are pending and pg_dump is not installed here: no backup is taken first \
                 (the Helm chart takes one in its pre-upgrade hook)"
            );
        }
    }
    let migrated = store.migrate_journaled(VERSION).await?;
    if migrated.from_schema != migrated.to_schema || migrated.resumed {
        tracing::info!(
            from = migrated.from_schema,
            to = migrated.to_schema,
            resumed = migrated.resumed,
            version = VERSION,
            "database upgraded"
        );
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn free_space_is_read_from_df() {
        let out = "Filesystem 1024-blocks Used Available Capacity Mounted on\n\
                   /dev/sda1 102400 2048 100352 2% /\n";
        assert_eq!(parse_df(out), Some(100_352 << 10));
        assert_eq!(parse_df(""), None);
        assert_eq!(parse_df("header only"), None);
        assert!(free_bytes(Path::new("/definitely/not/there/at/all")).is_some() || cfg!(windows));
    }
}

File: go/hub/internal/ops/upgrade/upgrade.go
// Package upgrade is upgrades (crates/kuben/src/cli/upgrade.rs, M4.8): the
// preflight `kuben upgrade-check` prints, run with the *new* binary before
// it replaces the old one, and the journaled migration every start and
// `kuben migrate` go through, with a backup first when migrations are
// pending. It imports nothing from internal/cli: the server calls it too.
package upgrade

import (
	"context"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/Teamtem-dev/kuben/go/hub/internal/core/clock"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/config"
	"github.com/Teamtem-dev/kuben/go/hub/internal/core/opt"
	coreupgrade "github.com/Teamtem-dev/kuben/go/hub/internal/core/upgrade"
	"github.com/Teamtem-dev/kuben/go/hub/internal/ops/backup"
	"github.com/Teamtem-dev/kuben/go/hub/internal/store"
	"github.com/Teamtem-dev/kuben/go/hub/internal/version"
	"github.com/Teamtem-dev/kuben/go/kubenapi/protocol"
)

// FreeBytes is the free space of the file system holding dir (its nearest
// existing ancestor), from `df`.
func FreeBytes(ctx context.Context, dir string) opt.Val[uint64] {
	existing := filepath.Clean(dir)
	for {
		if _, err := os.Stat(existing); err == nil {
			break
		}
		parent := filepath.Dir(existing)
		if parent == existing {
			return opt.None[uint64]()
		}
		existing = parent
	}
	out, err := exec.CommandContext(ctx, "df", "-Pk", existing).Output() //nolint:gosec // a fixed program on a path
	if err != nil {
		return opt.None[uint64]()
	}
	return ParseDF(string(out))
}

// ParseDF is the available KiB of the last line of `df -Pk`, in bytes.
func ParseDF(output string) opt.Val[uint64] {
	lines := strings.Split(strings.TrimRight(output, "\n"), "\n")
	fields := strings.Fields(lines[len(lines)-1])
	if len(fields) < 4 {
		return opt.None[uint64]()
	}
	kib, err := strconv.ParseUint(fields[3], 10, 64)
	if err != nil || kib > (1<<54)-1 {
		return opt.None[uint64]()
	}
	return opt.Some(kib << 10)
}

// facts is what the preflight of this binary against st sees.
func facts(ctx context.Context, cfg config.Config, st *store.Store, major bool) (coreupgrade.Facts, error) {
	schema, err := st.SchemaState(ctx)
	if err != nil {
		return coreupgrade.Facts{}, err //nolint:wrapcheck // a store error
	}
	versions, err := st.ServerVersions(ctx)
	if err != nil {
		return coreupgrade.Facts{}, err //nolint:wrapcheck // a store error
	}
	f := coreupgrade.Facts{
		Target:          version.Version,
		Major:           major,
		Schema:          schema.Applied,
		TargetSchema:    store.LatestMigration(),
		DirtySchema:     schema.Dirty,
		Now:             clock.System{}.NowMs(),
		TargetProtocols: coreupgrade.ProtocolRange{Min: protocol.SupportedVersions.Min, Max: protocol.SupportedVersions.Max},
		FreeBytes:       FreeBytes(ctx, cfg.BackupDir()),
	}
	if len(versions) > 0 {
		f.Running = opt.Some(versions[0])
	}
	if schema.Applied >= 25 {
		last, found, err := st.LastBackup(ctx)
		if err != nil {
			return coreupgrade.Facts{}, err //nolint:wrapcheck // a store error
		}
		if found {
			f.LastBackup = opt.Some(last.FinishedAt)
		}
	}
	if schema.Applied > 0 {
		if f.ActiveOperations, err = st.ActiveOperations(ctx); err != nil {
			return coreupgrade.Facts{}, err //nolint:wrapcheck // a store error
		}
		agents, err := st.AgentProtocols(ctx)
		if err != nil {
			return coreupgrade.Facts{}, err //nolint:wrapcheck // a store error
		}
		for _, a := range agents {
			f.Agents = append(f.Agents, coreupgrade.Agent{Cluster: a.Cluster, Protocol: a.Protocol})
		}
	}
	return f, nil
}

// Preflight is every finding of the preflight of this binary against the
// database of cfg, which it connects to without migrating.
func Preflight(ctx context.Context, cfg config.Config, major bool) ([]coreupgrade.Finding, error) {
	st, err := store.ConnectUnmigrated(ctx, cfg.Database)
	if err != nil {
		return nil, err //nolint:wrapcheck // explains itself
	}
	defer st.Close()
	f, err := facts(ctx, cfg, st, major)
	if err != nil {
		return nil, err
	}
	return coreupgrade.Preflight(f), nil
}

// Migrate migrates st as this version, journaled; when migrations are
// pending on a database with data, it takes a backup first where
// PostgreSQL's client tools are installed. A failed backup stops the
// migration.
func Migrate(ctx context.Context, cfg config.Config, st *store.Store, logger *slog.Logger) (store.Migrated, error) {
	state, err := st.GuardSchema(ctx)
	if err != nil {
		return store.Migrated{}, err //nolint:wrapcheck // a store error
	}
	pending := store.LatestMigration() - state.Applied
	if pending > 0 && state.Applied > 0 && cfg.Backup.BeforeUpgrade {
		if backup.ClientToolsInstalled() {
			logger.Info("taking a backup before the migrations", "pending", pending)
			if err := backup.Run(ctx, cfg, backup.Options{Scheduled: true}); err != nil {
				return store.Migrated{}, err //nolint:wrapcheck // explains itself
			}
		} else {
			logger.Warn("migrations are pending and pg_dump is not installed here: no backup is taken first "+
				"(the Helm chart takes one in its pre-upgrade hook)", "pending", pending)
		}
	}
	migrated, err := st.MigrateJournaled(ctx, version.Version)
	if err != nil {
		return store.Migrated{}, err //nolint:wrapcheck // a store error
	}
	if migrated.FromSchema != migrated.ToSchema || migrated.Resumed {
		logger.Info("database upgraded", "from", migrated.FromSchema, "to", migrated.ToSchema,
			"resumed", migrated.Resumed, "version", version.Version)
	}
	return migrated, nil
}
